### PR TITLE
State: Move gravatar status reducer out of current user reducer

### PR DESCRIPTION
**Pull request by tyxla**
#### Changes proposed in this Pull Request

Currently, much of Calypso's reducers are modular and are loaded on demand when needed. The only bits that form the monolithic reducer nowadays are:

* Current user
* Data layer
* Sites

As part of the monolithic reducer, these are always loaded on every Calypso page load.

The gravatar status reducer is part of the current user reducer, but it's not connected to it in any way. It's also used only in a couple of places - when displaying an avatar and when uploading a new one. 

So this PR splits out the gravatar status reducer out of the current user one. It also changes the gravatar status to be a modular one.

#### Testing instructions

* Verify Calypso builds well.
* Go to `/me`
* Verify you can still upload an avatar like before.
* Verify all tests pass.

### CODE REVIEW

File: client/state/gravatar-status/selectors.js [hunk 1]
```diff
@@ -17,7 +19,5 @@ export function isCurrentUserUploadingGravatar( state ) {
  * @returns {string|boolean} - The temp Gravatar string, or false
  */
 export function getUserTempGravatar( state, userId ) {
-	return (
-		state.currentUser.id === userId && ( state.currentUser.gravatarStatus.tempImage.src ?? false )
-	);
+	return state.currentUser.id === userId && ( state.gravatarStatus.tempImage.src ?? false );
```
Comment (jsnajdr) [inline client/state/gravatar-status/selectors.js]: `!! state.gravatarStatus.tempImage.src` is a bit more stylish 🙂 

Also, if you don't mind, the `tempImage` URL could be directly in `state.gravatarStatus.tempImage`. No need for the extra nested property `.src`.

Then the `state.currentUser.id` could be replaced with the `getCurrentUserId` selector which makes future `currentUser` refactors less error-prone.

**Review by jsnajdr — APPROVED** [review]
Looks good, but as always I have some suggestions 🙂

File: client/state/gravatar-status/selectors.js (hunk 1 above)
Comment (tyxla) [reply client/state/gravatar-status/selectors.js]: > `!! state.gravatarStatus.tempImage.src` is a bit more stylish 🙂

Agreed! Addressed in https://github.com/Automattic/wp-calypso/pull/62066/commits/1c9c9934cbcb3f52493af40bcce42c9b3045cf94.

> Also, if you don't mind, the `tempImage` URL could be directly in `state.gravatarStatus.tempImage`. No need for the extra nested property `.src`.

Happy to further simplify as a drive-by, done in https://github.com/Automattic/wp-calypso/pull/62066/commits/26e0c4c5dd9b68aa513d7ec850a715051e42155d

> Then the `state.currentUser.id` could be replaced with the `getCurrentUserId` selector which makes future `currentUser` refactors less error-prone.

Good call, updated in https://github.com/Automattic/wp-calypso/pull/62066/commits/11400e85d816875c6e6af20ce50760273598dcaf.

**Comment by tyxla** [comment]
@jsnajdr I've addressed your feedback; do you have a chance to take another look?

File: client/state/gravatar-status/selectors.js [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
+
 /**
  * Returns true if we're currently uploading a Gravatar
  *
  * @param {object} state - The state
  * @returns {boolean} - If uploading a Gravatar
  */
 export function isCurrentUserUploadingGravatar( state ) {
-	return state.currentUser.gravatarStatus.isUploading ?? false;
+	return state.gravatarStatus.isUploading ?? false;
```
Comment (jsnajdr) [inline client/state/gravatar-status/selectors.js]: Didn't notice this one before: the state is always a valid boolean so the `?? false` bit can be removed.

File: client/state/gravatar-status/test/reducer.js [hunk 3]
```diff
@@ -48,17 +48,15 @@ describe( 'reducer', () => {
 
 		test( 'returns empty object by default', () => {
 			const state = tempImage( undefined, {} );
-			expect( state ).to.eql( {} );
+			expect( state ).to.eql( null );
```
Comment (jsnajdr) [inline client/state/gravatar-status/test/reducer.js]: How much work would be to move the assertions from Chai to Jest?

**Review by jsnajdr — APPROVED** [review]

File: client/state/gravatar-status/selectors.js (hunk 1 above)
Comment (tyxla) [reply client/state/gravatar-status/selectors.js]: > > !! state.gravatarStatus.tempImage.src is a bit more stylish 🙂

> Agreed! Addressed in https://github.com/Automattic/wp-calypso/commit/1c9c9934cbcb3f52493af40bcce42c9b3045cf94.

Had to revert this, it was breaking the original logic of the selector.

File: client/state/gravatar-status/selectors.js (hunk 2 above)
Comment (tyxla) [reply client/state/gravatar-status/selectors.js]: Good call, done in https://github.com/Automattic/wp-calypso/pull/62066/commits/502f754397a23c6951f619f3d9287aa2367af282

File: client/state/gravatar-status/test/reducer.js (hunk 3 above)
Comment (tyxla) [reply client/state/gravatar-status/test/reducer.js]: Not much actually, done in https://github.com/Automattic/wp-calypso/pull/62066/commits/d0754411c728414a638df3e0772a87942d116f28.

**Comment by tyxla** [comment]
Did a bunch of changes as you suggested @jsnajdr so I'd appreciate a final look when you get a chance @jsnajdr 🙌

File: client/state/gravatar-status/test/reducer.js [hunk 4]
```diff
@@ -9,37 +8,38 @@ import reducer, { isUploading, tempImage } from '../reducer';
 
 describe( 'reducer', () => {
 	test( 'exports expected reducer keys', () => {
-		expect( reducer( undefined, {} ) ).to.have.keys( [ 'isUploading', 'tempImage' ] );
+		expect( reducer( undefined, {} ) ).toHaveProperty( 'isUploading' );
+		expect( reducer( undefined, {} ) ).toHaveProperty( 'tempImage' );
```
Comment (jsnajdr) [inline client/state/gravatar-status/test/reducer.js]: What about `expect( Object.keys( reducer() ) ).toEqual( [ 'isUploading', 'tempImage' ] )`?

**Review by jsnajdr — APPROVED** [review]
Still good 🙂

File: client/state/gravatar-status/test/reducer.js (hunk 4 above)
Comment (tyxla) [reply client/state/gravatar-status/test/reducer.js]: Interesting, but I think my version displays more transparently the intention of the test. I've improved it to call the reducer function just once in https://github.com/Automattic/wp-calypso/pull/62066/commits/30aae00303dfad9f235ae5141aef1f90eae5367e

**Comment by tyxla** [comment]
Thanks, @jsnajdr for another thorough review 🙌

**Pull request merged**